fed/tree_util.py

```diff
@@ -0,0 +1,339 @@
+    ...
+
+def map_only(ty: TypeAny) -> MapOnlyFn[FnAny[Any]]:
+    """
+    Suppose you are writing a tree_map over tensors, leaving everything
+    else unchanged.  Ordinarily you would have to write:
+
+        def go(t):
+            if isinstance(t, Tensor):
+                return ...
+            else:
+                return t
+
+    With this function, you only need to write:
+
+        @map_only(Tensor)
+        def go(t):
+            return ...
+
+    You can also directly use 'tree_map_only'
+    """
+    def deco(f: Callable[[T], Any]) -> Callable[[Any], Any]:
+        @functools.wraps(f)
+        def inner(x: T) -> Any:
+            if isinstance(x, ty):
+                return f(x)
+            else:
+                return x
+        return inner
+    return deco
+
+@overload
+def tree_map_only(ty: Type[T], fn: Fn[T, Any], pytree: PyTree) -> PyTree:
+    ...
+
+@overload
+def tree_map_only(ty: Type2[T, S], fn: Fn2[T, S, Any], pytree: PyTree) -> PyTree:
+    ...
+
+@overload
+def tree_map_only(ty: Type3[T, S, U], fn: Fn3[T, S, U, Any], pytree: PyTree) -> PyTree:
+    ...
+
+def tree_map_only(ty: TypeAny, fn: FnAny[Any], pytree: PyTree) -> PyTree:
+    return tree_map(map_only(ty)(fn), pytree)
+
+def tree_all(pred: Callable[[Any], bool], pytree: PyTree) -> bool:
+    flat_args, _ = tree_flatten(pytree)
+    return all(map(pred, flat_args))
+
+def tree_any(pred: Callable[[Any], bool], pytree: PyTree) -> bool:
+    flat_args, _ = tree_flatten(pytree)
+    return any(map(pred, flat_args))
+
+@overload
+def tree_all_only(ty: Type[T], pred: Fn[T, bool], pytree: PyTree) -> bool:
+    ...
+
+@overload
+def tree_all_only(ty: Type2[T, S], pred: Fn2[T, S, bool], pytree: PyTree) -> bool:
+    ...
+
+@overload
+def tree_all_only(ty: Type3[T, S, U], pred: Fn3[T, S, U, bool], pytree: PyTree) -> bool:
+    ...
+
+def tree_all_only(ty: TypeAny, pred: FnAny[bool], pytree: PyTree) -> bool:
+    flat_args, _ = tree_flatten(pytree)
+    return all(pred(x) for x in flat_args if isinstance(x, ty))
+
+@overload
+def tree_any_only(ty: Type[T], pred: Fn[T, bool], pytree: PyTree) -> bool:
+    ...
+
+@overload
+def tree_any_only(ty: Type2[T, S], pred: Fn2[T, S, bool], pytree: PyTree) -> bool:
+    ...
+
+def tree_any_only(ty: TypeAny, pred: FnAny[bool], pytree: PyTree) -> bool:
+    flat_args, _ = tree_flatten(pytree)
+    return any(pred(x) for x in flat_args if isinstance(x, ty))
+
+# Broadcasts a pytree to the provided TreeSpec and returns the flattened
+# values. If this is not possible, then this function returns None.
+#
+# For example, given pytree=0 and spec=TreeSpec(list, None, [LeafSpec(), LeafSpec()]),
+# would return [0, 0]. This is useful for part of the vmap implementation:
+# a user can pass in vmap(fn, in_dims)(*inputs). `in_dims` should be
+# broadcastable to the tree structure of `inputs` and we use
+# _broadcast_to_and_flatten to check this.
+def _broadcast_to_and_flatten(pytree: PyTree, spec: TreeSpec) -> Optional[List[Any]]:
```

Comment:
Any place is `_broadcast_to_and_flatten` used at ?

Reply:
Removed.